### DISCUSSION TITLE
OS Fractional Scaling interferes with mx-auto's spacing

### DISCUSSION THREAD

**Discussion by Neutrovertido** [tailwindlabs/tailwindcss#15039] [Help]
**What version of Tailwind CSS are you using?**

v3.4.15

**What build tool (or framework if it abstracts the build tool) are you using?**

Unsure about that, will list all dependencies:
"autoprefixer": "^10.4.20"
"postcss": "^8.4.49"
"tailwindcss": "^3.4.15"

I'm using ReactJS v13.4.0 as my framework too.

**What version of Node.js are you using?**

v22.11

**What browser are you using?**

Mozilla Firefox, but I confirmed it also happens in Chromium.

**What operating system are you using?**

Windows 11

**Reproduction URL**

Bug Report source code: https://github.com/Neutrovertido/tailwind-fractional-scaling-bug

Live preview: https://glowing-sunshine-667537.netlify.app/

**Describe your issue**

The issue seemingly appears only when using fractional scaling (ie. 125%, 150%) on the OS configuration and also only when tailwind uses the `xl` viewport.

This also seemingly doesn't happen when scaling using the web browser on 100% system scaling.

What happens is `mx-auto` considerably decreases the horizontal margin, making it look too close to the screen borders in comparison to 100% scaling.

To replicate this bug, simply put your "Scale" system setting to anything above 100% and load a site that uses `mx-auto` with a viewport of `xl`:
![image](https://github.com/user-attachments/assets/821765fb-a7ea-44e0-938b-11186cd841ec)

Then, compare it to how it looks on 100% scaling, and voila.

Expected:
![image](https://github.com/user-attachments/assets/2da2865c-7e5e-4932-9705-dc47f02bd1ff)

Result:
![image](https://github.com/user-attachments/assets/23802274-591e-48d4-95ba-78fc48cd39b7)

__This bug does _NOT_ happen on viewports other than `xl`__:
![image](https://github.com/user-attachments/assets/e4e92e83-0732-49cb-8a19-15f2fee6bbbb)

**Comment by adamwathan** [comment]
Hey! The `mx-auto` class just sets the following CSS:

```css
.mx-auto {
  margin-left: auto;
  margin-right: auto;
}
```

So any behavior you're seeing here is down to just a CSS thing — we're not doing anything clever here that we could do any differently. Whatever you're seeing here would happen if you wrote the CSS by hand as well even without Tailwind so I don't think there's any action we can take here.

Whatever is happening here can't really be a bug in Tailwind itself since there's no different CSS we could/should generate instead, so moving this to a discussion in case anyone is able to help you figure out the root cause of the behavior you're seeing.

**Reply by Neutrovertido** [reply]
I did fix it on my own project by doing some JavaScript to detect fractional scaling and adding extra padding if so.

But I just think it's weird for fractional scaling to break a website like that, be that because of tailwind or something else.

**Comment by Neutrovertido** [comment]
In case anyone stumbles upon this in the future, here's a potential (albeit not perfect) workaround:

```js

import React from "react";
import Header from "./components/Header";
import { useEffect } from 'react';

const App = () => {
  // Workaround for troublesome mx-auto not detecting fractional scaling
  useEffect(() => {
    function detectFractionalScaling() {
      const pixelRatio = window.devicePixelRatio;
      return pixelRatio >= 1.25 && pixelRatio % 1 !== 0; // If fractional scaling is on or above 125%
    }
    
    // This function will add a class that fixes the spacing with fractional scaling
    function adjustContainerForScaling() {
      const container = document.querySelector('.headercontainer');
      if (detectFractionalScaling() && container) {
        container.classList.add('px-16');
      } else if (container) {
        container.classList.remove('px-16');
      }
    }

    // These two functions will remove the spacing if the equivalent to XL viewport is not met
    function isXL() {
      return window.innerWidth >= 1280;
    }

    function resizeIfScaling() {
      const container = document.querySelector('.headercontainer');
      if (container.classList.contains('px-16') && !isXL()) {
        container.classList.remove('px-16');
        container.classList.add('px-4');
      } else if (container.classList.contains('px-4') && isXL()) {
        container.classList.remove('px-4');
        container.classList.add('px-16');
      }
    }

    // Run at website load
    adjustContainerForScaling();

    // Listen for resize events
    window.addEventListener('resize', resizeIfScaling);
    return () => window.removeEventListener('resize', resizeIfScaling);
  }, []);

  return (
    <div className="dark bg-gradient-to-tl from-slate-300 to-slate-500 text-neutral-200 min-h-screen">
      <div className="headercontainer container mx-auto p-4">
        <Header />
      </div>
    </div>
  );
};

export default App;

```
